mitodl/open-discussions
# Use react query for search page; fix bookmarking bug

#### Pre-Flight checklist


#### What are the relevant tickets?
#3908 

#### What's this PR do?
This PR changes `SearchPage` to use `react-query` to manage its infinite scroll. This is accomplished largely by upgrading  [`@mitodl/course-search-utils`](https://github.com/mitodl/course-search-utils)
 - Previously, we were using `useCourseSearch` course-search-utils. It managed search UI + some aspects of API requests.
 - Now we use `useSearchInputs` and `useSyncUrlAndSearch` (new exports of course-search-utils introduced in https://github.com/mitodl/course-search-utils/pull/51) to manage UI and URL syncing, and use react-query's `useInfiniteScroll` for managing the API queries.

Using `react-query` for managing the API calls has a few advantages: We get its builtin support for status info (loading, error, successful); we can let it handle storing the API results; if we ever wanted to switch to a paginated UI (as opposed to infinite scroll) that would be easy.

It also means that updates to lists / favorite status can appropriately patch the search results, which fixes #3908 

#### How should this be manually tested?
Because this includes a fairly large refactor of the search pag, we should test (1) all aspects of the search page, and (2) the bugfix for #3908.

**Test the search page:** Visit http://od.odl.local:8063/infinite/search and test things like:
- search results appear to be correct, and update when you scroll
- toggling a facet on/off automatically re-submits a query
- edit the search text and click submit; query should be correct
- check that the URL updates correctly when doing all of the above
- try the back button in your browser. Search UI (facets, text) & and search results should update
- check the network tab in dev console...Are extra search requests being made? _Should be "no"._

**Test bookmark status is updated appropriately:** (bug #3908)
As an authenticated user who has some userlists:
1. visit http://od.odl.local:8063/infinite/search
2. Click the bookmark button to add/remove result from a list. The UI should update quickly and without a reload.
3. Also add/remove item from "Favorites" (which is in the list modal, but not really a list). The UI should update quickly and without a reload.

#### Where should the reviewer start?
Start by looking at the new `SearchPage.tsx` file to see how the updated course-search-utils is being used. 

### Related
https://github.com/mitodl/course-search-utils/pull/51
